swap/swap.go

```diff
@@ -408,27 +410,14 @@ func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitCheque
 		return err
 	}
 
-	otherSwap, err := contract.InstanceAt(cheque.Contract, s.backend)
+	expectedPayout, transactionCosts, err := s.cashoutProcessor.estimatePayout(context.TODO(), cheque)
 	if err != nil {
-		log.Error("error getting contract", "err", err)
 		return err
 	}
 
-	gasPrice, err := s.backend.SuggestGasPrice(context.TODO())
-	if err != nil {
-		return err
-	}
-	transactionCosts := gasPrice.Uint64() * 50000 // cashing a cheque is approximately 50000 gas
-	paidOut, err := otherSwap.PaidOut(nil, cheque.Beneficiary)
-	if err != nil {
-		return err
-	}
 	// do a payout transaction if we get 2 times the gas costs
-	if (cheque.CumulativePayout - paidOut.Uint64()) > 2*transactionCosts {
-		opts := bind.NewKeyedTransactor(s.owner.privateKey)
-		opts.Context = ctx
-		// cash cheque in async, otherwise this blocks here until the TX is mined
-		go defaultCashCheque(s, otherSwap, opts, cheque)
+	if expectedPayout > 2*transactionCosts {
+		go defaultCashCheque(s, cheque)
```

Comment:
Probably I asked before, so please be patient with me :), but if we initiate a `cashCheque` then for some reason a new `handleEmitChequeMsg` runs, is there any concurrency issue? I guess it's not possible to cash the same cheque twice, but any other risks?

Reply:
Hmm. Good question!
To me, it seems that this will happen:

Assume tx costs are 20
- Cheque comes in with an estimated payout of 41 => we attempt cashing
- Right after, same cheque with a new new cumulative payment comes in. Estimated payout is now 42 => we attempt cashing.

End result is that we cash a cheque with a worth of 21 and a cheque with a worth of 1. We have a net loss on this, as 1 from the seconds transaction is less than the tx costs of 20. 

In my opinion, the problem really lies in the `estimatePayout` function not taking into account any pending cashouts.

Reply:
This is only a temporary problem. Once concurrent cashings are processed sequentially this won't happen anymore. (The idea being that `estimatePayout` happens when the actual cashing attempt happens, not when it is queued).

Reply:
Is this the queuing concept? Do we actually have an issue for that?

Reply:
yes #2006.

Reply:
Arguably not the most detailed issue description. I will add more details to the issue soon.